File: src/services/ingestion-service/Pipeline/ModeVocabulary.cs
namespace Traverse.IngestionService.Pipeline;

/// <summary>
/// The mode vocabulary the CPLM engine understands, mirrored here so ingestion can
/// tell — at the moment it maps a MODE — whether the value it is about to publish
/// will mean anything downstream.
///
/// Why this exists: the engine's <c>isAutoMode()</c> is binary. Anything it does not
/// recognise is simply "not auto", so a wrong or missing <c>mode_value_map</c> does not
/// error — it silently marks every sample manual and G1 excludes the whole window.
/// That is exactly what happened on the HDPE plant (2026-09-09): the map named CENTUM
/// value 4 as AUT while 1 was the controlling mode, so the entire fleet was excluded
/// while G0 stayed green and the data looked perfectly healthy. Nothing in the chain
/// said a word. This class is the missing word.
///
/// Keep in sync with CplmNormalizedSample.AUTO_MODE_TOKENS / MANUAL_MODE_TOKENS
/// (src/flink/src/main/java/com/ams/flink/cplm/CplmNormalizedSample.java).
/// </summary>
public static class ModeVocabulary
{
    private static readonly HashSet<string> Auto = new(StringComparer.Ordinal)
    {
        "AUTO", "AUT", "A", "AUTOMATIC", "NORMAL", "NORM",
        "CAS", "CASC", "CASCADE", "RSP", "DDC", "SUP", "SUPERVISORY",
    };

    private static readonly HashSet<string> Manual = new(StringComparer.Ordinal)
    {
        "MAN", "MANUAL", "M", "IMAN", "ROUT", "LO", "LOCAL", "OFF", "TRACK",
    };

    /// <summary>
    /// False when the engine would fall through to its substring fallback and treat the
    /// value as not-auto by default rather than by recognition — i.e. the loop will be
    /// excluded and nobody will be told why.
    /// </summary>
    public static bool IsRecognised(string? mode)
    {
        if (string.IsNullOrWhiteSpace(mode)) return false;
        var m = mode.Trim().ToUpperInvariant();
        if (m.Length == 0 || m == "UNKNOWN") return false;
        if (Manual.Contains(m) || Auto.Contains(m)) return true;
        // The engine's own compound-vendor-string fallback; recognised, if loosely.
        return m.Contains("AUTO", StringComparison.Ordinal)
            || m.Contains("CASCADE", StringComparison.Ordinal);
    }

    /// <summary>True when the engine will count this mode as closed-loop control.</summary>
    public static bool IsAuto(string? mode)
    {
        if (string.IsNullOrWhiteSpace(mode)) return false;
        var m = mode.Trim().ToUpperInvariant();
        if (m.Length == 0 || m == "UNKNOWN") return false;
        if (Manual.Contains(m)) return false;          // manual wins, as in the engine
        if (Auto.Contains(m)) return true;
        return m.Contains("AUTO", StringComparison.Ordinal)
            || m.Contains("CASCADE", StringComparison.Ordinal);
    }
}

File: src/frontend-ob/src/utils/modeVocabulary.ts
/**
 * The controller-mode vocabulary the CPLM engine understands.
 *
 * Mirror of ingestion-service `Pipeline/ModeVocabulary.cs`, itself a mirror of Flink
 * `CplmNormalizedSample.AUTO_MODE_TOKENS / MANUAL_MODE_TOKENS`. Keep the three in step.
 *
 * Why it matters on screen: the engine's auto test is binary, so anything it does not
 * recognise is simply "not auto" — the loop is excluded at Gate 1 and produces no verdict
 * at all. On the HDPE plant a wrong `mode_value_map` hid that for weeks: G0 stayed green,
 * the data looked perfect, and every loop was silently unscored. Showing the mode, and
 * what it means for analysis, is how that stays visible.
 */

const AUTO = new Set([
  'AUTO', 'AUT', 'A', 'AUTOMATIC', 'NORMAL', 'NORM',
  'CAS', 'CASC', 'CASCADE', 'RSP', 'DDC', 'SUP', 'SUPERVISORY',
]);
const MANUAL = new Set([
  'MAN', 'MANUAL', 'M', 'IMAN', 'ROUT', 'LO', 'LOCAL', 'OFF', 'TRACK',
]);

export type ModeClass = 'auto' | 'manual' | 'unrecognised';

/** What the engine will make of this token. Manual wins before auto, as in the engine. */
export function classifyMode(token: string): ModeClass {
  const m = token.trim().toUpperCase();
  if (!m || m === 'UNKNOWN') return 'unrecognised';
  if (MANUAL.has(m)) return 'manual';
  if (AUTO.has(m)) return 'auto';
  if (m.includes('AUTO') || m.includes('CASCADE')) return 'auto';
  return 'unrecognised';
}

export type ModeTone = 'good' | 'warn' | 'bad' | 'muted';

/**
 * Display form for a live MODE value: the token, plus what it means for whether this
 * loop is scored at all. `UNKNOWN` is its own case — it means the source has never
 * published a mode, not that it sent something we could not read.
 */
export function modeLabel(value: unknown): { label: string; note: string; tone: ModeTone } {
  if (value === undefined || value === null || value === '') {
    return { label: '—', note: 'no live value, none stored', tone: 'muted' };
  }
  const token = String(value).trim().toUpperCase();
  if (token === 'UNKNOWN') {
    return { label: 'UNKNOWN', note: 'mode never published · excluded at G1', tone: 'warn' };
  }
  switch (classifyMode(token)) {
    case 'auto':
      return { label: token, note: 'closed loop · analysed', tone: 'good' };
    case 'manual':
      return { label: token, note: 'manual · excluded at G1', tone: 'warn' };
    default:
      return { label: token, note: 'not recognised · excluded at G1', tone: 'bad' };
  }
}
